Update BDD Markdown to v.6

pudkrong opened the pull request:
The BDD Markdown v.6 introduces breaking change.

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- DeviceInfo
- DeviceToWebsocket

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- DeviceShadow

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- HistoricalData

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- LastSeen

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- MultipleNRFCloudAccounts

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- LocationTrail

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- SingleCellGeoLocation

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- HistoricalRequest

pudkrong committed:
test: refactor to use BDD markdown v. 6

## Features
- DeviceConfiguration
- RuntimeConfiguration

pudkrong committed:
test: clean up

pudkrong committed:
chore: update dependencies

pudkrong committed:
fix: bdd markdown API change

pudkrong committed:
chore: update dependencies

pudkrong committed:
test: using generated request id in testing

pudkrong committed:
Merge branch 'saga' into refactor-e2e

pudkrong committed:
test: adjust retry to match the task

pudkrong committed:
fix: downgrade swc

pudkrong committed:
test: adjust retry time for shadow

coderbyheart commented on feature-runner/run-features.ts:
```diff
@@ -89,33 +89,13 @@ const runner = await runFolder({
-				chalk.magenta(info.context.title),
-				...args.map((arg) => chalk.cyan(print(arg))),
-			),
+			console.error(ts(), ...args.map((arg) => chalk.cyan(print(arg)))),
```
Unrelated color change. Why?

pudkrong replied:
I just remove `chalk.magenta(info.context.title)` because `info.context` is not available since BDD markdown v.6

coderbyheart replied:
Ah, I see. `context` has been renamed to `step`, it now is only the instance of the Step.

coderbyheart commented on feature-runner/steps/config.ts:
```diff
@@ -18,36 +15,26 @@ const createConfigStepRunners = ({
-	}: StepRunnerArgs<Record<string, any>>): Promise<StepRunResult> => {
-		const match = matchGroups(
-			Type.Object({
+	const setupDeviceShadowFetchingConfiguration2 = regExpMatchedStep(
```
```suggestion
	const setupDeviceShadowFetchingConfiguration = regExpMatchedStep(
```

coderbyheart commented on feature-runner/steps/device.ts:
```diff
@@ -21,190 +18,178 @@ import { getAttributesForDevice } from '../../devices/getAttributesForDevice.js'
-			Type.Object({
+	)
+
+const getDevice = ({
```
Remove unused step definition.

coderbyheart commented on features/HistoricalRequest.feature.md:
```diff
@@ -49,7 +56,7 @@ And I send this message via the websocket
   "message": "message",
   "payload": {
     "@context": "https://github.com/hello-nrfcloud/proto/historical-data-request",
-    "@id": "420eac59-5ce8-4751-b7d1-217811382095",
+    "@id": "${requestId}",
```
Unrelated change. What's wrong with using a hardcoded ID here?

pudkrong replied:
When I debug the freaky test, I would like to have a unique request id. So, it is easier to search on the log.

coderbyheart commented on features/LastSeen.feature.md:
```diff
@@ -23,15 +31,12 @@ And the device `${fingerprint_deviceId}` publishes this message to the topic
-When I connect to the websocket using fingerprint `${fingerprint}`
-
-<!-- @retry:tries=5,initialDelay=5000,delayFactor=1 -->
+When I connect to the websocket using fingerprint `${fingerprint}` in `5`
+seconds
```
```suggestion
Given I wait for 5 seconds

When I connect to the websocket using fingerprint `${fingerprint}`
```

If user must wait 5 seconds before connect, add a separate wait step.

coderbyheart replied:
But I wonder, why must the user wait now?

pudkrong replied:
Last seen will be recorded in DB in async way. So, sometimes when you disconnect and reconnect immediately, the DB is not done yet and the test is failed.

coderbyheart commented on package.json:
```diff
@@ -44,22 +44,22 @@
     "@nordicsemiconductor/cloudformation-helpers": "8.0.0",
     "@nordicsemiconductor/eslint-config-asset-tracker-cloud-typescript": "16.0.18",
-    "@nordicsemiconductor/firmware-ci-device-helpers": "14.0.198",
+    "@nordicsemiconductor/firmware-ci-device-helpers": "14.0.201",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -44,22 +44,22 @@
+    "@nordicsemiconductor/firmware-ci-device-helpers": "14.0.201",
     "@swc/cli": "0.1.62",
-    "@swc/core": "1.3.84",
+    "@swc/core": "1.3.83",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -44,22 +44,22 @@
+    "@swc/core": "1.3.83",
     "@swc/jest": "0.2.29",
-    "@types/aws-lambda": "8.10.119",
+    "@types/aws-lambda": "8.10.120",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -44,22 +44,22 @@
     "@types/ws": "8.5.5",
     "@types/yazl": "2.4.2",
-    "@typescript-eslint/eslint-plugin": "6.6.0",
+    "@typescript-eslint/eslint-plugin": "6.7.0",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -73,14 +73,14 @@
     "glob": "10.3.4",
     "husky": "8.0.3",
     "id128": "1.6.6",
-    "jest": "29.6.4",
+    "jest": "29.7.0",
     "jsonata": "2.0.3",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -73,14 +73,14 @@
     "semver": "7.5.4",
     "tsmatchers": "4.0.12",
-    "tsx": "3.12.8",
+    "tsx": "3.12.9",
```
Unrelated change. Revert.

coderbyheart commented on package.json:
```diff
@@ -118,11 +118,11 @@
   "dependencies": {
     "@aws-lambda-powertools/metrics": "1.12.1",
-    "@hello.nrfcloud.com/proto": "5.5.4",
+    "@hello.nrfcloud.com/proto": "5.5.10",
```
Unrelated change. Revert.

pudkrong replied:
We need to upgrade `@hello.nrfcloud.com/proto` to use the same version of `@sinclair/typebox`

coderbyheart commented on package.json:
```diff
@@ -118,11 +118,11 @@
     "@nordicsemiconductor/from-env": "2.0.0",
     "@nordicsemiconductor/timestream-helpers": "5.0.0",
-    "@sinclair/typebox": "0.31.5",
+    "@sinclair/typebox": "0.31.14",
```
Unrelated change. Revert.

pudkrong replied:
We need the same version of `@sinclair/typebox` in `@nordicsemiconductor/bdd-markdown`

pudkrong committed:
test: refactor code

pudkrong committed:
Revert "chore: update dependencies"

This reverts commit 51df96f8338f5b9eca59e55e12d8bc8619ce6f92.

pudkrong committed:
Revert "fix: downgrade swc"

This reverts commit be4cc065245a93d4b8b8a66e165daee12237ab6a.

pudkrong committed:
test: using hardcoded ID

pudkrong committed:
fix: emit assert for import assertions

as per https://github.com/hello-nrfcloud/proto/commit/2c50f8afa5af066c07c125e9851be9b2b68675fa

pudkrong committed:
test: revert unrelated changes

coderbyheart commented on feature-runner/steps/utils.ts:
Remove.

coderbyheart commented on features/RuntimeConfiguration.feature.md:
```diff
@@ -281,18 +287,16 @@ And there is this device shadow data for `${fingerprint_deviceId}` in nRF Cloud
 Given device shadow fetching config for model `PCA20035+solar` is `10`
 
-When I connect to the websocket using fingerprint `${fingerprint}`
+When I wait for `5` seconds
```
This is nothing the user should do, the 5 second delay should be handled in the following connect step. It's an implementation detail of the test system and should not be made part of the public API.

pudkrong replied:
But if we put the logic in web socket connection, it will apply for all even we need it for a specific case.

coderbyheart replied:
Yes, it's the DB that is lacking behind so this needs an internal retry for the connection db.

coderbyheart replied:
And this check should be in the promise of the step that **provides** the data in the db. So this step waits until it is available and only then do the next steps execute. If the step is later called, the promise is already fulfilled and it will continue immediately.

coderbyheart commented on package.json:
```diff
@@ -118,11 +118,11 @@
   "dependencies": {
     "@aws-lambda-powertools/metrics": "1.12.1",
-    "@hello.nrfcloud.com/proto": "5.5.4",
+    "@hello.nrfcloud.com/proto": "5.5.10",
```
Unrelated change, revert.

pudkrong replied:
This change is required to satisfy `tsc` because all `typebox` modules must be the same version.

coderbyheart commented on package.json:
```diff
@@ -118,11 +118,11 @@
     "@nordicsemiconductor/from-env": "2.0.0",
     "@nordicsemiconductor/timestream-helpers": "5.0.0",
-    "@sinclair/typebox": "0.31.5",
+    "@sinclair/typebox": "0.31.14",
```
Unrelated change, revert.

pudkrong replied:
This change is required to satisfy tsc because all typebox modules must be the same version.

pudkrong committed:
test: remove unrelated steps for user

coderbyheart commented on feature-runner/run-features.ts:
```diff
@@ -89,33 +89,13 @@ const runner = await runFolder({
-				chalk.magenta(info.context.title),
-				...args.map((arg) => chalk.cyan(print(arg))),
-			),
+			console.error(ts(), ...args.map((arg) => chalk.cyan(print(arg)))),
```
```suggestion
			console.error(
				ts(),
				chalk.magenta.dim(info.step.keyword),
				chalk.magenta(info.step.title),
				...args.map((arg) => chalk.cyan(print(arg))),
			),
```

pudkrong committed:
fix: add more info on the log

pudkrong committed:
fix: using package-lock.json from linux
